File: leetcode-js/143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if (!head || !head.next) {
        return;
    }

    let fast = head;
    let slow = head;

    while (fast.next && fast.next.next) {
        fast = fast.next.next;
        slow = slow.next;
    }

    let middle = reverseList(slow.next);
    slow.next = null;

    const dummy = new ListNode(0);
    let node = dummy;

    while (middle) {
        node.next = head;
        head = head.next;
        node.next.next = middle;
        middle = middle.next;
        node = node.next.next;
    }

    node.next = head;
};

function reverseList(node) {
    let prev = null;

    while (node) {
        const next = node.next;
        node.next = prev;
        prev = node;
        node = next;
    }

    return prev;
}

File: leetcode-java/143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        if (head == null) {
            return;
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode mid = slow.next;

        slow.next = null;

        ListNode node1 = head;
        ListNode node2 = reverse(mid);
        ListNode dummy = new ListNode(0);
        ListNode node = dummy;

        while (node1 != null) {
            node.next = node1;
            node = node.next;
            node1 = node1.next;
            if (node2 != null) {
                node.next = node2;
                node = node.next;
                node2 = node2.next;
            }
        }
    }

    ListNode reverse(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode dummy = new ListNode(0);
        ListNode cur = head;

        dummy.next = head;

        while (cur.next != null) {
            ListNode start = dummy.next;
            dummy.next = cur.next;
            cur.next = dummy.next.next;
            dummy.next.next = start;
        }

        return dummy.next;
    }
}
